src/audio/component.c

```diff
@@ -328,162 +328,286 @@ int comp_verify_params(struct comp_dev *dev, uint32_t flag,
 	return 0;
 }
 
-void comp_free_model_data(struct comp_dev *dev, struct comp_model_data *model)
+/** \brief Struct handler for large component configs */
+struct comp_data_blob_handler {
+	uint32_t data_size;	/**< size of component's data blob */
+	void *data;		/**< pointer to data blob */
+	void *data_new;		/**< pointer to new data blob */
+	bool data_ready;	/**< set when data blob is fully received */
+	uint32_t data_pos;	/**< indicates a data position in data
+				  *  sending/receiving process
+				  */
+};
+
+static void comp_free_data_blob(struct comp_dev *dev,
+				struct comp_data_blob_handler *blob_handler)
 {
-	if (!model->data)
+	if (!blob_handler || !blob_handler->data)
 		return;
 
-	rfree(model->data);
-	model->data = NULL;
-	model->data_size = 0;
-	model->crc = 0;
-	model->data_pos = 0;
+	rfree(blob_handler->data);
+	rfree(blob_handler->data_new);
+	blob_handler->data = NULL;
+	blob_handler->data_new = NULL;
+	blob_handler->data_size = 0;
 }
 
-int  comp_alloc_model_data(struct comp_dev *dev, struct comp_model_data *model,
-			   uint32_t size)
+void *comp_get_data_blob(struct comp_dev *dev,
+			 struct comp_data_blob_handler *blob_handler,
+			 size_t *size)
 {
-	comp_free_model_data(dev, model);
+	comp_dbg(dev, "comp_get_data_blob()");
+
+	if (size)
+		*size = 0;
+
+	if (!blob_handler) {
+		comp_err(dev, "comp_get_data_blob(): !blob_handler");
+		return NULL;
+	}
+
+	/* Function returns new data blob if available */
+	if (comp_is_new_data_blob_available(dev, blob_handler)) {
+		comp_dbg(dev, "comp_get_data_blob(): new data available");
+
+		/* Free "old" data blob and set data to data_new pointer */
+		rfree(blob_handler->data);
+		blob_handler->data = blob_handler->data_new;
+		blob_handler->data_new = NULL;
+		blob_handler->data_ready = false;
+
+		if (size)
+			*size = blob_handler->data_size;
+
+		return blob_handler->data;
+	}
+
+	/* If blob_handler->data is equal to NULL and there is no new data blob
+	 * it means that component hasn't got any config yet. Function returns
+	 * NULL in that case.
+	 */
+	if (!blob_handler->data) {
```

Comment:
This if() should just trace warning and continue to avoid another return.

Reply:
@mmaka1 
Updated.